Posts/2021/02/XML-NewEmployees.md

```diff
@@ -0,0 +1,180 @@
+```
+Dead simple. 4 questions with 4 answers, into 4 variables. Instead of asking lots of questions to find out information like postcode or country, let's use our XML!
+
+```powershell
+# Declare some more variables
+$FirstInitial = "$FirstName.substring(0,1)"
+$Path = 'OU=Users,OU=Contoso,DC=Contoso,DC=com'
+
+# Build hashtable of new user parameters
+  $NewUserParams =@{
+    
+        GivenName = $FirstName
+        Name = "$FirstName $LastName"
+        Surname = $LastName
+        SamAccountName = "$FirstInitial + $LastName"
+        DisplayName = "$FirstName $LastName"
+        UserPrincipalName = "$userName" + "@" + "contoso.com"
+        AccountPassword = $password
+        Office = $Office
+        StreetAddress = ($XML.Companies.$Company.$office.Street).Trim()
+        PostalCode = ($XML.Companies.$Company.$Office.Postcode).Trim()
+        City = ($XML.Companies.$Company.$Office.City).Trim()
+        State = ($xml.Companies.$Company.$Office.State).Trim()
+        Country = ($xml.Companies.$Company.$Office.Country).Trim()
+        Path = $Path
+        ChangePasswordAtLogon = $false
+        Enabled = $true
+    }
+    
+# Create the user
+New-ADUser @NewUserParams
+```
+Simple as that, browsing through the xml elements using dot-notation, and trimming off any unwanted whitespace. This is cropped below for convenience:
+
+```powershell
+
+        StreetAddress = ($XML.Companies.$Company.$office.Street).Trim()
+        PostalCode = ($XML.Companies.$Company.$Office.Postcode).Trim()
+        City = ($XML.Companies.$Company.$Office.City).Trim()
+        State = ($xml.Companies.$Company.$Office.State).Trim()
+        Country = ($xml.Companies.$Company.$Office.Country).Trim()
+```
+# Helpful Notes
+* Be careful when creating your elements! No whitespace allowed!
+* Your inputted variables using Read-Host need to match the references in the XML. You could make use of Validation Scripts to validate your inputs. Here is a helpful read on everything to do with validating inputs: https://adamtheautomator.com/powershell-validatescript/
+
+# Final Comments and Credits
+
+You can really flesh this out more and get a load of information into XML for referencing, e.g. company project information, sets of AD groups users could be added to, managed bookmark URLS (to then be copied into registry items). 
+
+The list goes on and on: they are just strings at the end of the day! 
+
+I was inspired by the following people to both do this work and write this blog post.
+
+* Thomas Lee (https://twitter.com/doctordns) - thanks for getting this blog started in the first place!
+* Adam Bertram (https://twitter.com/adbertram) and his fantastic book "PowerShell for SysAdmins: Workflow Automation Made Easy" (https://nostarch.com/powershellsysadmins)
+* Tobias Weltner (*I couldn't find your twitter :(*) and his article on Mastering everday XML Tasks in PowerShell (https://www.powershellmagazine.com/2013/08/19/mastering-everyday-xml-tasks-in-powershell/)
+
```

Comment:
restyle the URLs as noted above.
Also, Tobias's Twitter handle is **@TobiasPSP** (and no relation to my company PS Partnership !).